### PR TITLE
Code Browser Added

### PR DESCRIPTION
**Pull request by julian-garcia-hernandez**
Hey guys, here is my code browser.

Going to list what I contributed in this PR as well as acknowledge what I still need to do.
Note: I am just preemptively submitting this PR with stuff that works so far. I am still going to work on the "What I didn't do" section tonight and see if I could whip up something.
What I did:
- Added a `CodeBrowser` component that takes in `language` and `code` (both string types) as props.
- The `CodeBrowser`  component supports syntax-highlighting with the usage of the prism.js library
- The `CodeBrowser` component manifests when clicking on the `SubmissionCard`. To make it dissapear, simply click on the `SubmissionCard` component again. If there are design choices that could be better on this, please feel free to make a suggestion.

What I didn't do:
- Request the correct file from the Supabase bucket depending on which student's `SubmissionCard` I just clicked. Currently all `SubmissionCard` clicks result in the same file being fed to the `BrowserComponent`. Because of this, you will notice the same `5_hw1.py` file be generated upon clicking on any one of the submission cards in the Assignment View of OpenGrader. I intend to figure out a way to fix this as soon as possible.
- Instructor comments.

### CODE REVIEW

File: Components/CodeBrowser.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  language: string;
+  code: string;
+}
+
+const CodeBrowser: React.FC<CodeBrowserProps> = ({ language, code }) => {
+  if (typeof code !== "string"){
+    return <div>Value is in wrong format</div>
+  } else if (typeof code === 'undefined' || typeof code === null){
+    return <div>Value is undefined or null</div>
+  }
```
Comment (jacksonwelsh) [inline Components/CodeBrowser.tsx:15]: this branch will never evaluate

File: pages/course/[course]/assignment/[assignment].tsx [hunk 2]
```diff
@@ -130,7 +154,24 @@ const AssignmentView: NextPage<AssignmentProps> = ({ assignment }) => {
         </h1>
         <p>{assignment.description}</p>
         <h2 className="font-semibold text-2xl text-slate-50">Submissions</h2>
-        {assignment.submission.map(SubmissionCard)}
+          {assignment.submission.map(submission => {
+            if (typeof submission.student === "number") {
+              return null;
+            }
```
Comment (jacksonwelsh) [inline pages/course/[course]/assignment/[assignment].tsx:160]: how does this happen?

File: package.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
     "format:fix": "prettier --write --ignore-path .gitignore ."
   },
   "dependencies": {
+    "@fontsource/fira-code": "^4.5.13",
```
Comment (jacksonwelsh) [inline package.json:11]: remove in lieu of google fonts, you can just update the @​import we already have in css

File: util/misc.ts [hunk 4]
```diff
@@ -38,4 +38,4 @@ export const getCurrentUser = async (client: SupabaseClient): Promise<User | nul
     alerts: { missing: 0, plagiarism: 0, errors: 0 },
     auth_id: profile.auth_id,
   };
-};
+};
```
Comment (jacksonwelsh) [inline util/misc.ts:41]: let's drop this change

**Review by jacksonwelsh — CHANGES_REQUESTED** [review]
minor formatting things and a couple notes

File: pages/course/[course]/assignment/[assignment].tsx (hunk 2 above)
Comment (julian-garcia-hernandez) [reply pages/course/[course]/assignment/[assignment].tsx]: I was confused on that myself. Didn't know `submission.student` could be `User | number | null` type. The compiler kept complaining if I didn't do some error checking.

File: package.json (hunk 3 above)
Comment (julian-garcia-hernandez) [reply package.json:11]: Okay going to remove.

File: util/misc.ts (hunk 4 above)
Comment (julian-garcia-hernandez) [reply util/misc.ts]: Okay going to drop

Comment (julian-garcia-hernandez) [reply util/misc.ts]: I assume this gets fixed with prettier?

File: pages/course/[course]/assignment/[assignment].tsx [hunk 5]
```diff
@@ -28,6 +33,15 @@ export const getServerSideProps = (ctx: GetServerSidePropsContext) =>
     const courseId = queryParamToNumber(ctx.query?.course);
 
     const supabase = createServerSupabaseClient(ctx);
+
+    let blob;
+    let code;
+    blob = await supabaseAdmin.storage.from("assignments").download(`8/5_hw1.py`);  //hardcoded; going to fix ASAP;
+    
+    if (blob != null && blob.data != null){
```
[the remark this answers is not in the document]
Comment (julian-garcia-hernandez) [reply pages/course/[course]/assignment/[assignment].tsx]: Okay going to shorten

File: Components/CodeBrowser.tsx (hunk 1 above)
Comment (julian-garcia-hernandez) [reply Components/CodeBrowser.tsx]: You're right, I was doing some error checking, thought it was necessary. Getting rid of it.

Comment (julian-garcia-hernandez) [reply Components/CodeBrowser.tsx]: I just ran prettier

File: util/misc.ts (hunk 4 above)
Comment (jacksonwelsh) [reply util/misc.ts]: yeah it's fixed now

File: package.json (hunk 3 above)
Comment (jacksonwelsh) [reply package.json]: you'll need to run `pnpm i` again to get CI to pass as well

**Comment by jacksonwelsh** [comment]
idk why but this has a bunch of old changes and all the commits got duplicated... don't think i'll be able to review this tonight beyond just taking your word for it

**Review by ekobedevon — COMMENTED** [review]
Good besides supabase server client question

File: pages/course/[course]/assignment/[assignment].tsx [hunk 6]
```diff
@@ -28,6 +33,15 @@ export const getServerSideProps = (ctx: GetServerSidePropsContext) =>
     const courseId = queryParamToNumber(ctx.query?.course);
 
     const supabase = createServerSupabaseClient(ctx);
+
+    let blob;
+    let code;
+    blob = await supabaseAdmin.storage.from("assignments").download(`8/5_hw1.py`); //hardcoded; going to fix ASAP;
```
[the remark this answers is not in the document]
Comment (julian-garcia-hernandez) [reply pages/course/[course]/assignment/[assignment].tsx:39]: I tried using the regular one and it gave me an error, I want to say @jacksonwelsh knows more about this. It has something to do with roles?

[the remark this answers is not in the document]
Comment (ekobedevon) [reply pages/course/[course]/assignment/[assignment].tsx:39]: heard

**Review by ekobedevon — APPROVED** [review]

File: pages/course/[course]/assignment/[assignment].tsx (hunk 2 above)
Comment (jacksonwelsh) [reply pages/course/[course]/assignment/[assignment].tsx]: yeah i'd say dw about it for now, we can use the server client and rls later

File: pages/course/[course]/assignment/[assignment].tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (jacksonwelsh) [reply pages/course/[course]/assignment/[assignment].tsx:39]: yeah i'd say dw about it for now, we can use the server client and rls later

**Review by jacksonwelsh — APPROVED** [review]

**Pull request merged**